<a href="https://colab.research.google.com/github/3132Ganesh/wtsapp_group_dna/blob/main/ganeshnayak_mini_project_1_Group_DNA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

wtsapp_group_DNA


In [ ]:
from datetime import datetime

def parse_chat(filename):
    chat_data = []
    with open(filename, "r", encoding="utf-8") as f:
        for line in f:

            if " - " in line:
                try:
                    timestamp_str, rest = line.split(" - ", 1)

                    dt = datetime.strptime(timestamp_str.strip(), "%d/%m/%y, %H:%M")
                    if ": " in rest:
                        sender, message = rest.split(": ", 1)
                        chat_data.append({
                            "timestamp": dt,
                            "sender": sender.strip(),
                            "message": message.strip(),
                            "type": "text"
                        })
                    else:

                        chat_data.append({
                            "timestamp": dt,
                            "sender": None,
                            "message": rest.strip(),
                            "type": "system"
                        })
                except ValueError:

                    continue
    return chat_data

parsed = parse_chat("hostel_bois.txt")
for entry in parsed[:10]:  # Shows first 10 entries
    print(entry)

{'timestamp': datetime.datetime(2024, 4, 1, 1, 12), 'sender': None, 'message': 'Messages and calls are end-to-end encrypted. No one outside of this chat, not even WhatsApp, can read or listen to them. Tap to learn more.', 'type': 'system'}
{'timestamp': datetime.datetime(2024, 4, 1, 1, 13), 'sender': None, 'message': 'Priya created group "Hostel Bois 4ever"', 'type': 'system'}
{'timestamp': datetime.datetime(2024, 4, 1, 1, 14), 'sender': None, 'message': 'Priya added you', 'type': 'system'}
{'timestamp': datetime.datetime(2024, 4, 1, 1, 17), 'sender': 'Rahul', 'message': 'scene fix', 'type': 'text'}
{'timestamp': datetime.datetime(2024, 4, 1, 1, 17), 'sender': 'Rahul', 'message': 'haan', 'type': 'text'}
{'timestamp': datetime.datetime(2024, 4, 1, 1, 18), 'sender': 'Rahul', 'message': 'kya scene', 'type': 'text'}
{'timestamp': datetime.datetime(2024, 4, 1, 2, 13), 'sender': 'Rahul', 'message': 'abhi free hai?', 'type': 'text'}
{'timestamp': datetime.datetime(2024, 4, 1, 2, 13), 'sender'

In [ ]:
def grp_overview(chat_data):
  total_mssg=len(chat_data)
  participants=set()
  per_person_count={}
  for e in chat_data:
    sender=e['sender']
    if sender:
      participants.add(sender)
      per_person_count[sender]=per_person_count.get(sender,0)+1
  start_D=chat_data[0]["timestamp"]
  end_D=chat_data[-1]["timestamp"]
  print("===================================================\n GROUP OVERVIEW \n===================================================")
  print(f"total mssgs :{total_mssg}")
  print(f" participants :{len(participants)}->\nNAMES:{','.join(participants)}")
  print(f"start date  :{start_D}  \nEnd date  :{end_D}\n")
  print("MESSAGE PER PERSON")
  for sender,count in per_person_count.items():
    print(f"{sender} : {count}")
  return participants


In [ ]:
participants = grp_overview(parsed)

 GROUP OVERVIEW 
total mssgs :3178
 participants :6->
NAMES:Aman,Neha,Vikas,Priya,Rahul,Karan
start date  :2024-04-01 01:12:00  
End date  :2024-05-30 23:31:00

MESSAGE PER PERSON
Rahul : 953
Priya : 718
Karan : 354
Neha : 635
Aman : 490
Vikas : 24


In [ ]:
from collections import Counter
from datetime import datetime
def most_active_day_hr(chat_data):
  dayscount=Counter()
  hrscount=Counter()
  for e in chat_data:
    ts = e['timestamp']
    dayscount[ts.date()]+=1
    hrscount[ts.hour]+=1

  if not dayscount:
      print("No valid timestamps found to determine active days/hours.")
      return

  busiest_day=max(dayscount,key=dayscount.get)
  busiest_hr=max(hrscount,key=hrscount.get)
  print("Most Active Day & hour")
  print(f"Busiest day:{busiest_day}({dayscount[busiest_day]} message)")

  print(f"Busiest hour :{busiest_hr} :00 ({hrscount[busiest_hr]} messages)")
most_active_day_hr(parsed)

Most Active Day & hour
Busiest day:2024-05-04(76 message)
Busiest hour :18 :00 (248 messages)


TOP 5 BUSIES DAYS AND HRS


In [ ]:
from collections import Counter
from datetime import datetime

def top_5_days_hours(chat_data):
    day_counts = Counter()
    hour_counts = Counter()

    for entry in chat_data:
        ts = entry["timestamp"] # ts is now guaranteed to be a datetime object
        # Removed datetime.strptime, as it's no longer needed and would cause a TypeError
        day_counts[ts.date()] += 1
        hour_counts[ts.hour] += 1

    # Top 5 busiest days
    top_days = day_counts.most_common(5)
    # Top 5 busiest hours
    top_hours = hour_counts.most_common(5)

    print(" Top 5 Busiest Days")
    print("Date         | Messages")
    print("-----------------------")
    for day, count in top_days:
        print(f"{day} | {count}")

    print("\n Top 5 Busiest Hours")
    print("Hour   | Messages")
    print("------------------")
    for hour, count in top_hours:
        print(f"{hour:02d}:00 | {count}")
top_5_days_hours(parsed)

 Top 5 Busiest Days
Date         | Messages
-----------------------
2024-05-04 | 76
2024-04-24 | 72
2024-04-28 | 72
2024-04-03 | 71
2024-04-27 | 65

 Top 5 Busiest Hours
Hour   | Messages
------------------
18:00 | 248
19:00 | 228
12:00 | 193
16:00 | 189
21:00 | 177


ACTIVITY HEATMAP


In [ ]:
import numpy as np
from datetime import datetime
import numpy as np

def activity_heatmap(chat_data):
    participants = sorted({e["sender"] for e in chat_data if e["sender"]})
    heatmap = np.zeros((len(participants), 24), dtype=int)

    for e in chat_data:
        sender = e["sender"]
        ts = e["timestamp"]
        if sender and isinstance(ts, datetime):
            hour = ts.hour
            row = participants.index(sender)
            heatmap[row, hour] += 1

    print("ACTIVITY HEATMAP (messages by hour)")
    print("Hour:       " + " ".join([f"{h:02d}" for h in range(24)]))
    for i, sender in enumerate(participants):
        row_data = " ".join([str(x).rjust(2) for x in heatmap[i]])
        print(f"{sender:10}| {row_data}")
    return heatmap, participants

heatmap, participants_from_heatmap = activity_heatmap(parsed)

ACTIVITY HEATMAP (messages by hour)
Hour:       00 01 02 03 04 05 06 07 08 09 10 11 12 13 14 15 16 17 18 19 20 21 22 23
Aman      | 54 67 66 60 88  0  0  0  0  0  0  0  0  0 14 11 19  7 16  8 13 11  0 56
Karan     |  0  0  0  0  0  0  0  4 12 16 20 16 37 25 32 27 27 27 25 32 23 14  9  8
Neha      |  0  0  0  0  0 19  3 13 36 52 52 22 39 36 27 10 37 47 62 50 45 27 28 30
Priya     |  0  0  0  0  0  0 13 20 47 65 62 61 57 48 44 29 32 40 38 60 43 32 18  9
Rahul     |  3 15 17 17 22 10 17 17 24 17 25 15 58 48 45 53 73 49 105 76 41 92 60 54
Vikas     |  0  0  0  0  0  0  0  1  3  1  1  0  2  2  0  1  1  3  2  2  1  1  1  2


In [ ]:
import numpy as np
from datetime import datetime

def activity_heatmap(chat_data):
    participants = sorted({e["sender"] for e in chat_data if e["sender"]})
    heatmap = np.zeros((len(participants), 24), dtype=int)

    for e in chat_data:
        sender = e["sender"]
        ts = e["timestamp"]
        if sender:

            hour = ts.hour
            row = participants.index(sender)
            heatmap[row, hour] += 1

    print("ACTIVITY HEATMAP (messages by hour)")
    print("Hour:      " + " ".join([f"{h:02d}" for h in range(24)]))

    for i, sender in enumerate(participants):
        max_val = heatmap[i].max()
        row_visual = []
        for val in heatmap[i]:
            if max_val == 0:
                row_visual.append(". ")
            else:
                ratio = val / max_val
                if ratio == 0:
                    row_visual.append("  ")
                elif ratio <= 0.25:
                    row_visual.append("░  ")
                elif ratio <= 0.50:
                    row_visual.append("▒  ")
                elif ratio <= 0.75:
                    row_visual.append("▓  ")
                else:
                    row_visual.append("█  ")
        print(f"{sender:10}| {''.join(row_visual)}")
activity_heatmap(parsed)

ACTIVITY HEATMAP (messages by hour)
Hour:      00 01 02 03 04 05 06 07 08 09 10 11 12 13 14 15 16 17 18 19 20 21 22 23
Aman      | ▓  █  ▓  ▓  █                    ░  ░  ░  ░  ░  ░  ░  ░    ▓  
Karan     |               ░  ▒  ▒  ▓  ▒  █  ▓  █  ▓  ▓  ▓  ▓  █  ▓  ▒  ░  ░  
Neha      |           ▒  ░  ░  ▓  █  █  ▒  ▓  ▓  ▒  ░  ▓  █  █  █  ▓  ▒  ▒  ▒  
Priya     |             ░  ▒  ▓  █  █  █  █  ▓  ▓  ▒  ▒  ▓  ▓  █  ▓  ▒  ▒  ░  
Rahul     | ░  ░  ░  ░  ░  ░  ░  ░  ░  ░  ░  ░  ▓  ▒  ▒  ▓  ▓  ▒  █  ▓  ▒  █  ▓  ▓  
Vikas     |               ▒  █  ▒  ▒    ▓  ▓    ▒  ▒  █  ▓  ▓  ▒  ▒  ▒  ▓  


In [ ]:
import string
def top_words(chat_data,top_n=10):
  stop_words = {"i","is","the","a","and","or","to","of","in","on","for","you","me","it","my","we","was","this","how","about","at","am","so","he","she","his","just","which","today","that","up","have","one","are","from","had","they","no"}
  word_count={}
  for e in chat_data:
    word=e["message"].lower().split()
    for w in word:
      w=w.strip(string.punctuation)
      if w and w not in stop_words:
        word_count[w]=word_count.get(w,0)+1
  sorted_words=sorted(word_count.items(),key=lambda x:x[1],reverse=True)[:top_n]
  print("THIS GROUP'S FRVT WORDS")
  for word,count in sorted_words:
    bar="‖"*(count//5)
    print(f"{word:<10} {bar} {count}")
  return word_count
word_count = top_words(parsed)

THIS GROUP'S FRVT WORDS
guys       ‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖ 318
hai        ‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖ 268
everyone   ‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖ 187
telling    ‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖ 179
bhai       ‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖ 160
started    ‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖ 150
scene      ‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖ 145
entire     ‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖ 145
please     ‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖ 141
anyone     ‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖‖ 139


In [ ]:
def response_stats(chat_data):
    response_times = {}
    silent_streaks = {}

    prev_time = None
    prev_sender = None

    for entry in chat_data:
        sender = entry["sender"]
        ts = entry["timestamp"]

        if sender and isinstance(ts, datetime):
            # Response time calculation
            if prev_time and sender != prev_sender:
                diff = (ts - prev_time).total_seconds() / 60  # minutes
                response_times.setdefault(sender, []).append(diff)

            # Silent streak calculation
            if prev_time:
                diff = (ts - prev_time).total_seconds() / 60
                silent_streaks[sender] = max(silent_streaks.get(sender, 0), diff)

            prev_time = ts
            prev_sender = sender

    print(" Response Speed (avg minutes)")
    avg_response_times = {}
    for sender, times in response_times.items():
        avg = sum(times) / len(times)
        avg_response_times[sender] = avg
        print(f"{sender:10} | {avg:.2f} min")

    if avg_response_times:
        fastest_replier = min(avg_response_times, key=avg_response_times.get)
        slowest_replier = max(avg_response_times, key=avg_response_times.get)
        print(f"\nFastest Replier: {fastest_replier} ({avg_response_times[fastest_replier]:.2f} min)")
        print(f"Slowest Replier: {slowest_replier} ({avg_response_times[slowest_replier]:.2f} min)")

    print("\n Longest Silent Streak")
    for sender, streak in silent_streaks.items():
        if streak >= 1440:  # more than a day
            days = streak / 1440
            print(f"{sender:10} | {days:.1f} days")
        elif streak >= 60:  # more than an hour
            hours = streak / 60
            print(f"{sender:10} | {hours:.1f} hours")
        else:
            print(f"{sender:10} | {streak:.1f} min")
    return response_times, silent_streaks
response_times, silent_streaks = response_stats(parsed)

 Response Speed (avg minutes)
Priya      | 41.99 min
Rahul      | 34.95 min
Karan      | 36.62 min
Neha       | 39.45 min
Aman       | 55.36 min
Vikas      | 46.30 min

Fastest Replier: Rahul (34.95 min)
Slowest Replier: Aman (55.36 min)

 Longest Silent Streak
Rahul      | 3.5 hours
Priya      | 6.0 hours
Karan      | 4.8 hours
Neha       | 5.4 hours
Aman       | 5.4 hours
Vikas      | 5.3 hours


feature 7:archtype Detection Logic

In [ ]:
def detect_archetypes(chat_data, word_counts, response_times, silent_streaks, heatmap, participants):
    archetypes = {}

    # Spammer: highest message count
    msg_counts = {p: sum(heatmap[i]) for i, p in enumerate(participants)}
    spammer = max(msg_counts, key=msg_counts.get)
    archetypes[spammer] = "Spammer"

    # Ghost: longest silent streak
    ghost = max(silent_streaks, key=silent_streaks.get)
    archetypes[ghost] = "Ghost"

    # Group Mom/Dad: most caring words
    caring_words = {"please", "everyone", "guys", "anyone"}
    caring_scores = {p: sum(word_counts.get(w, 0) for w in caring_words) for p in participants}
    mom = max(caring_scores, key=caring_scores.get)
    archetypes[mom] = "Group Mom/Dad"

    # Night Owl: most messages after 10
    night_scores = {p: sum(heatmap[i][22:]) for i, p in enumerate(participants)}
    night_owl = max(night_scores, key=night_scores.get)
    archetypes[night_owl] = "Night Owl"

    # Early Bird: most messages before 08
    early_scores = {p: sum(heatmap[i][:8]) for i, p in enumerate(participants)}
    early_bird = max(early_scores, key=early_scores.get)
    archetypes[early_bird] = "Early Bird"

    # Print results
    print(" Group Archetypes")
    for p in participants:
        role = archetypes.get(p, "Balanced Contributor")
        print(f"{p:10} → {role}")
detect_archetypes(parsed, word_count, response_times, silent_streaks, heatmap, participants_from_heatmap)

 Group Archetypes
Aman       → Early Bird
Karan      → Balanced Contributor
Neha       → Balanced Contributor
Priya      → Ghost
Rahul      → Night Owl
Vikas      → Balanced Contributor


In [ ]:
from collections import Counter


def conversation_flow(chat_data):
    # Count messages per day
    day_counts = Counter([e["timestamp"].date() for e in chat_data if e["sender"]])

    # Sort by date
    dates = sorted(day_counts.keys())
    counts = [day_counts[d] for d in dates]

    # Print top peaks
    print("📈 Conversation Peaks")
    for d, c in sorted(day_counts.items(), key=lambda x: x[1], reverse=True)[:5]:
        print(f"{d} → {c} messages")

    # Plot flow


conversation_flow(parsed)

📈 Conversation Peaks
2024-05-04 → 76 messages
2024-04-24 → 72 messages
2024-04-28 → 72 messages
2024-04-03 → 71 messages
2024-04-27 → 65 messages


In [ ]:
participants = grp_overview(parsed)
print("===========================================================\n")
most_active_day_hr(parsed)
print("============================================================\n")
top_5_days_hours(parsed)
print("============================================================\n")
activity_heatmap(parsed)
print("============================================================\n")
word_count = top_words(parsed)
print("=============================================================\n")
response_times, silent_streaks = response_stats(parsed)
print("=============================================================\n")
detect_archetypes(parsed, word_count, response_times, silent_streaks, heatmap, participants_from_heatmap)
print("==============================================================\n")
conversation_flow(parsed)

 GROUP OVERVIEW 
total mssgs :3178
 participants :6->
NAMES:Aman,Neha,Vikas,Priya,Rahul,Karan
start date  :2024-04-01 01:12:00  
End date  :2024-05-30 23:31:00

MESSAGE PER PERSON
Rahul : 953
Priya : 718
Karan : 354
Neha : 635
Aman : 490
Vikas : 24

Most Active Day & hour
Busiest day:2024-05-04(76 message)
Busiest hour :18 :00 (248 messages)

 Top 5 Busiest Days
Date         | Messages
-----------------------
2024-05-04 | 76
2024-04-24 | 72
2024-04-28 | 72
2024-04-03 | 71
2024-04-27 | 65

 Top 5 Busiest Hours
Hour   | Messages
------------------
18:00 | 248
19:00 | 228
12:00 | 193
16:00 | 189
21:00 | 177

ACTIVITY HEATMAP (messages by hour)
Hour:      00 01 02 03 04 05 06 07 08 09 10 11 12 13 14 15 16 17 18 19 20 21 22 23
Aman      | ▓  █  ▓  ▓  █                    ░  ░  ░  ░  ░  ░  ░  ░    ▓  
Karan     |               ░  ▒  ▒  ▓  ▒  █  ▓  █  ▓  ▓  ▓  ▓  █  ▓  ▒  ░  ░  
Neha      |           ▒  ░  ░  ▓  █  █  ▒  ▓  ▓  ▒  ░  ▓  █  █  █  ▓  ▒  ▒  ▒  
Priya     |             ░  ▒  ▓  █ 

## Final Report: Hostel Bois Chat Analysis

This report summarizes the key insights derived from the analysis of the `hostel_bois.txt` WhatsApp chat data, covering group activity, communication patterns, and participant archetypes.

### 1. Group Overview
*   **Total Messages:** 3178
*   **Participants:** 6 (Karan, Rahul, Vikas, Neha, Priya, Aman)
*   **Chat Duration:** From 2024-04-01 01:12:00 to 2024-05-30 23:31:00

**Message Distribution Per Person:**
*   Rahul : 953
*   Priya : 718
*   Karan : 354
*   Neha : 635
*   Aman : 490
*   Vikas : 24

### 2. Activity Peaks
*   **Busiest Day:** 2024-05-04 with 76 messages.
*   **Busiest Hour:** 18:00 (6 PM) with 248 messages.

**Top 5 Busiest Days:**
*   2024-05-04 | 76 messages
*   2024-04-24 | 72 messages
*   2024-04-28 | 72 messages
*   2024-04-03 | 71 messages
*   2024-04-27 | 65 messages

**Top 5 Busiest Hours:**
*   18:00 | 248 messages
*   19:00 | 228 messages
*   12:00 | 193 messages
*   16:00 | 189 messages
*   21:00 | 177 messages

### 3. Frequent Words
The most frequently used words (excluding common stop words) in the chat are:
*   `guys` (318 times)
*   `hai` (268 times)
*   `everyone` (187 times)
*   `telling` (179 times)
*   `bhai` (160 times)

### 4. Response Statistics
*   **Fastest Replier:** Rahul (average 34.95 minutes)
*   **Slowest Replier:** Aman (average 55.36 minutes)

**Longest Silent Streaks:**
*   Rahul: 3.5 hours
*   Priya: 6.0 hours
*   Karan: 4.8 hours
*   Neha: 5.4 hours
*   Aman: 5.4 hours
*   Vikas: 5.3 hours

### 5. Participant Archetypes
Based on message patterns, response times, silent streaks, and active hours, the following archetypes were detected:
*   **Aman:** Early Bird (most active in the morning)
*   **Karan:** Balanced Contributor
*   **Neha:** Balanced Contributor
*   **Priya:** Ghost (longest silent streak)
*   **Rahul:** Night Owl (most active in late hours)
*   **Vikas:** Balanced Contributor

### 6. Conversation Flow
**Top 5 Conversation Peaks (by message count per day):**
*   2024-05-04 → 76 messages
*   2024-04-24 → 72 messages
*   2024-04-28 → 72 messages
*   2024-04-03 → 71 messages
*   2024-04-27 → 65 messages

This analysis provides a comprehensive overview of the `Hostel Bois 4ever` group's communication dynamics.

# New Section